src/crypto/nss.rs

```diff
@@ -29,6 +34,118 @@ impl From<nss_gk_api::Error> for CryptoError {
 
 pub type Result<T> = std::result::Result<T, CryptoError>;
 
+// DER encode a pair of 32 byte unsigned integers as an RFC 3279 Ecdsa-Sig-Value.
+//   Ecdsa-Sig-Value  ::=  SEQUENCE  {
+//        r     INTEGER,
+//        s     INTEGER }.
+fn encode_der_p256_sig(r: &[u8], s: &[u8]) -> Result<Vec<u8>> {
+    if r.len() != 32 || s.len() != 32 {
+        return Err(CryptoError::MalformedInput);
+    }
+    // Each of the inputs is no more than 32 bytes as an unsigned integer. So each is no more than
+    // 33 bytes as a signed integer and no more than 35 bytes with tag and length.  The surrounding
+    // tag and length for the SEQUENCE has length 2, so the output is no more than 72 bytes.
+    let mut out = Vec::with_capacity(72);
+    out.push(DER_TAG_SEQUENCE);
+    out.push(0xaa); // placeholder for final length
+
+    let encode_u256 = |out: &mut Vec<u8>, r: &[u8]| {
+        // trim leading zeros, leaving a single zero if the input is the zero vector.
+        let mut r = r;
+        while r.len() > 1 && r[0] == 0 {
+            r = &r[1..];
+        }
+        out.push(DER_TAG_INTEGER);
+        if r[0] & 0x80 != 0 {
+            // Pad with a zero byte to avoid r being interpreted as a negative value.
+            out.push((r.len() + 1) as u8);
+            out.push(0x00);
+        } else {
+            out.push(r.len() as u8);
+        }
+        out.extend_from_slice(r);
+    };
+
+    encode_u256(&mut out, r);
+    encode_u256(&mut out, s);
+
+    // Write the length of the sequence
+    out[1] = (out.len() - 2) as u8;
+
+    Ok(out)
+}
+
+// Given "tag || len || value || rest" where tag and len are of length one, len is in [0, 127],
+// and value is of length len, returns (value, rest)
+#[cfg(test)]
+fn der_expect_tag_with_short_len(tag: u8, z: &[u8]) -> Result<(&[u8], &[u8])> {
+    if z.is_empty() {
+        return Err(CryptoError::MalformedInput);
+    }
+    let (h, z) = z.split_at(1);
+    if h[0] != tag || z.is_empty() {
+        return Err(CryptoError::MalformedInput);
+    }
+    let (h, z) = z.split_at(1);
+    if z.len() >= 0x80 || h[0] as usize > z.len() {
```

Comment:
Why is it important to enforce that `z.len() < 0x80`? `z` includes `value` and `rest`, so this puts a constraint on `rest` as well, which seems outside the scope of this function.

Reply:
good catch!